### PR TITLE
Fix multi line code block rendering in Jira subscription post

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
#### Summary
Fix multi-line code block rendering in the Jira subscription post

#### Old
Screenshot from 2025-01-27 19-33-06

#### New
Screenshot from 2025-01-27 19-33-11

#### Testing steps
- Create a subscription for JIra
- Comment on the Jira issue with multi-line code block with and without specifying language

### CODE REVIEW

File: server/webhook_parser.go [hunk 1]
```diff
@@ -405,16 +400,74 @@ func preProcessText(jiraMarkdownString string) string {
...
+		lines := strings.Split(content, "\n")
+		for i := range lines {
+			if lines[i] != "" {
+				lines[i] = "`" + lines[i] + "`"
+			}
+		}
+
+		if len(lines) == 1 {
+			return "`" + lines[0] + "`"
+		}
```
Comment (manojmalik20) [inline server/webhook_parser.go:433]: This case also should've been covered in the loop above.

File: server/webhook_parser.go [hunk 2]
```diff
@@ -405,16 +400,74 @@ func preProcessText(jiraMarkdownString string) string {
...
+			}
+		}
+
+		if len(lines) > 1 {
+			lines[len(lines)-1] = "`" + strings.TrimSuffix(lines[len(lines)-1], "{noformat}") + "`"
+		}
+
+		if len(lines) == 1 {
+			return "`" + lines[0] + "`"
+		}
```
Comment (manojmalik20) [inline server/webhook_parser.go:466]: This should've been covered by lines 450-452.

**Review by manojmalik20 — APPROVED** [review]

File: server/webhook_parser.go (hunk 2 above)
Comment (Kshitij-Katiyar) [reply server/webhook_parser.go:466]: Required condition to check to make sure no `\n` is added to single line of block code.
can move this code block below line 448 if we want

Comment (manojmalik20) [reply server/webhook_parser.go:466]: Okay, move this to below line 448 and make the same change in the first comment if possible.

Comment (Kshitij-Katiyar) [reply server/webhook_parser.go:466]: Possible, doing it now

**Review by manojmalik20 — APPROVED** [review]

File: server/webhook_parser.go [hunk 3]
```diff
@@ -405,16 +400,75 @@ func preProcessText(jiraMarkdownString string) string {
...
-		content := noFormatBlock[strings.Index(noFormatBlock, "}")+1 : strings.LastIndex(noFormatBlock, "{noformat}")]
-		return fmt.Sprintf("`%s`", content)
-	})
-
 	processedString = doubleCurlyRegex.ReplaceAllStringFunc(processedString, func(match string) string {
 		content := match[2 : len(match)-2]
 		return fmt.Sprintf("`%s`", content)
 	})
 
+	// handles sinlge and multi line language specific code blocks
```
Comment (wiggin77) [inline server/webhook_parser.go:408]: Typo in comment.

**Comment by wiggin77** [comment]
We need unit tests covering these cases. Can be a separate ticket.

**Comment by Kshitij-Katiyar** [comment]
> We need unit tests covering these cases. Can be a separate ticket. I'll approve once the ticket is created.

@wiggin77 Added them in this PR, please have a look.

**Review by wiggin77 — APPROVED** [review]

File: server/webhook_parser.go (hunk 3 above)
Comment (Kshitij-Katiyar) [reply server/webhook_parser.go]: Fixed

**Comment by Kshitij-Katiyar** [comment]
> While testing this PR, I found an issue in which the code snippet are not found working. Though, the code snippets are working fine for multiple linnes, but if incase any line is kept empty in between, the code(in comment) is shown as a normal text in Mattermost.

@AayushChaudhary0001 This issue has been fixed, please re test

**Comment by AayushChaudhary0001** [comment]
On re-testing this PR, the original issue looks to be fixed. However, a few new issues were found which are as follows:-

1. If the space between two lines of code is 20 empty lines or 50 empty lines or a single line, MM will show it as a single line only in every case.
2. If inside the code snippet, the total number of backticks are even, then code will be shown in multiple lines, but if the number of backticks are odd, then the code will be ahown as a normal text with no line break.
3. If the code is written inside backticks(after applying the code snippet), it consistd of a single backtick at the end of every word written.

**Screenshots for reference:-**

For case 1:-
image

image

For case 2:-
image

image

For case 3:-
image

image



Note:- Since, it looks like we are doing custom handling of the scenarios, the possible number of edge cases will be high.

**Comment by Kshitij-Katiyar** [comment]
> On re-testing this PR, the original issue looks to be fixed. However, a few new issues were found which are as follows:-
> 
> 1. If the space between two lines of code is 20 empty lines or 50 empty lines or a single line, MM will show it as a single line only in every case.
> 2. If inside the code snippet, the total number of backticks are even, then code will be shown in multiple lines, but if the number of backticks are odd, then the code will be ahown as a normal text with no line break.
> 3. If the code is written inside backticks(after applying the code snippet), it consistd of a single backtick at the end of every word written.
> 
> **Screenshots for reference:-**
> 
> For case 1:- image
> 
> image
> 
> For case 2:- image
> 
> image
> 
> For case 3:- image
> 
> image
> 
> Note:- Since, it looks like we are doing custom handling of the scenarios, the possible number of edge cases will be high.

@AayushChaudhary0001 The edge cases for custom handling for markdown are never-ending. We have covered the major and widely used markdowns supported by Mattermost in this and the previous #1151 PR.
We should test this PR extensively, note down all the markdown issues, and log them in a new Issue ticket so that we can merge this PR for now.

**Comment by arush-vashishtha** [comment]
This PR has been tested for the intended fix regarding the multi-line code blocks. Nevertheless, upon extensive testing we have found a few more cases related to the markdowns that are not working properly and we have logged those issues on Github for [supported markdowns](https://github.com/mattermost/mattermost-plugin-jira/issues/1159) and [unsupported markdowns](https://github.com/mattermost/mattermost-plugin-jira/issues/1160).

**Review by arush-vashishtha — APPROVED** [review]
